File: java/Solution215.java
package leetCode.java;

//Binary Search O(n) O(logn), 快排時間複雜度為nlogn, 此處只排單邊, 空間複雜度為遞迴期望值
class Solution215 {
    public int findKthLargest(int[] nums, int k) {
        // Can you solve it without sorting? -> quickSelect
        return quickSelect(0, nums.length - 1, nums, k - 1); // 記得轉index, 原k是第幾個
    }

    public int quickSelect(int start, int end, int[] nums, int index) {
        int pivot = nums[start];
        int l = start + 1;
        int r = end;

        // 由大到小排列
        while (l <= r) {
            if (nums[l] > pivot) {
                ++l;
            } else if (nums[r] < pivot) {
                --r;
            } else {
                swap(l++, r--, nums); // 記得要移動
            }
        }

        swap(start, r, nums);

        if (index == r) {
            return nums[r];
        } else if (index < r) {
            return quickSelect(start, r - 1, nums, index);
        } else { // index > r
            return quickSelect(r + 1, end, nums, index);
        }
    }

    public void swap(int a, int b, int[] nums) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}
/* 大排到小 k=2
3 4 5 1 2
p l     r

3 4 5 1 2
p   r l   r交換pivot

5 4 3 1 2
--- - ---

5 4
p l
  r

5 4
p l
r   r交換pivot

5 4
- -

4
p l
r   r交換pivot

4
-   return 4
*/


//Heap O(nlogn) O(k), n = nums.size()
class Solution215_2 {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(); // min heap, 小排到大
        for (int num : nums) {
            pq.offer(num);
            if (pq.size() > k) { // 拿掉最小的
                pq.poll();
            }
        }
        return pq.peek();
    }
}

File: c++/Solution215.cpp
//Binary Search O(n) O(logn)
class Solution215 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelect(nums, 0, nums.size() - 1, k - 1); // k-1轉為位置
    }

    int quickSelect(vector<int>& nums, int start, int end, int k) {
        int pivot = nums[start];
        int l = start + 1;
        int r = end;
        while (l <= r) {
            // 二路快排, 等於pivot也不跳過, 這樣遇到中間有大量相同數的test case, 左右partition會比較一致
            if (nums[l] > pivot) {
                ++l;
            } else if (nums[r] < pivot){
                --r;
            } else {
                swap(nums[l++], nums[r--]);
            }
        }
        swap(nums[start], nums[r]); // 此時r會較大, 移到左邊

        if (k == r) {
            return nums[r];
        } else if (k < r) {
            return quickSelect(nums, start, r - 1, k);
        } else { // k > l
            return quickSelect(nums, r + 1, end, k);
        }
    }
};


//Heap O(nlogn) O(k), n = nums.size()
class Solution215_2 {
public:
    int findKthLargest(vector<int>& nums, int k) { // 用pq會花比較多時間
        priority_queue<int, vector<int>, greater<int>> pq; // 小到大, min heap
        for (int num : nums) {
            pq.push(num);
            if (pq.size() > k) {
                pq.pop();
            }
        }
        return pq.top();
    }
};